### PR TITLE
[Feature] UnitedDeployment support patch for pool

### PR DESCRIPTION
**Pull request by kadisi**
this pr let unitedDeployment support patch for pool.

ref [issue 242](https://github.com/openyurtio/openyurt/issues/242)

for Pool Struct , add new field `Patches` :
```
// Pool defines the detail of a pool.
type Pool struct {
	// Indicates pool name as a DNS_LABEL, which will be used to generate
	// pool workload name prefix in the format '<deployment-name>-<pool-name>-'.
	// Name should be unique between all of the pools under one UnitedDeployment.
	// Name is NodePool Name
	Name string `json:"name"`

	// Indicates the node selector to form the pool. Depending on the node selector,
	// pods provisioned could be distributed across multiple groups of nodes.
	// A pool's nodeSelectorTerm is not allowed to be updated.
	// +optional
	NodeSelectorTerm corev1.NodeSelectorTerm `json:"nodeSelectorTerm,omitempty"`

	// Indicates the tolerations the pods under this pool have.
	// A pool's tolerations is not allowed to be updated.
	// +optional
	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`

	// Indicates the number of the pod to be created under this pool.
	// +required
	Replicas *int32 `json:"replicas,omitempty"`

	// Indicates the patches for the templateSpec
	// Now support strategic merge path :https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch
	// +optional
	Patches *runtime.RawExtension `json:"patches,omitempty"`
}
```

User can use this `Patches` to apply  `workloadTemplate.DeploymentTemplate` or `workloadTemplate.StatefulSetTemplate` configuration. and now only support strategic merge path.

**For the creation process**, workload inherits the `workloadTemplate` configuration. If `Patches` are found in the Pool, it applies the `workloadTemplate` using the strategic merge path. It also saves the patches configuration in the annotation. Please review these documents in detail:

pkg/yurtappmanager/controller/uniteddeployment/adapter/deployment_adapter.go `ApplyPoolTemplate`
pkg/yurtappmanager/controller/uniteddeployment/adapter/statefulset_adapter.go `ApplyPoolTemplate`

**For the update process**, the controller determines whether the annotation patches configuration in the workload matches the pool patches configuration. If not, trigger an update operation.

this is an example:
```
apiVersion: apps.openyurt.io/v1alpha1
kind: UnitedDeployment
metadata:
  name: deploy
  namespace: default
spec:
  selector:
    matchLabels:
      app: deploy
  workloadTemplate:
    deploymentTemplate:
      metadata:
        labels:
          app: deploy
      spec:
        selector:
          matchLabels:
            app: deploy
        template:
          metadata:
            labels:
              app: deploy
          spec:
            containers:
            - image: nginx:1.19.3
              imagePullPolicy: Always
              name: nginx
  topology:
    pools:
    - name: cloud
      nodeSelectorTerm:
        matchExpressions:
        - key: apps.openyurt.io/uniteddeployment
          operator: In
          values:
          - cloud
      patch:
        spec:
          template:
            spec:
              containers:
              - image: nginx:1.18.0
                name: nginx
      replicas: 1
```
add an `patch` field for pool

cc @Fei-Guo @huangyuqi @charleszheng44

### CODE REVIEW

**Comment by rambohe-ch** [comment]
@kadisi Would you make up an example that uses the patch field?

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/deployment_adapter.go [hunk 1]
```diff
@@ -136,6 +137,32 @@ func (a *DeploymentAdapter) ApplyPoolTemplate(ud *alpha1.UnitedDeployment, poolN
...
+	patched := &appsv1.Deployment{}
+	if err := StrategicMergeByPatches(set, poolConfig.Patches, patched); err != nil {
+		klog.Errorf("Deployment[%s/%s-] strategic merge by patch %s error %v", set.Namespace,
+			set.GenerateName, string(poolConfig.Patches.Raw), err)
+		return err
+	}
+	klog.Infof("Deployment [%s/%s-] has patches configure successfully:%v", set.Namespace,
+		set.GenerateName, string(poolConfig.Patches.Raw))
+
+	*set = *patched
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/adapter/deployment_adapter.go:158]: I think you need to use DeepCopy method here.

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/statefulset_adapter.go [hunk 2]
```diff
@@ -148,6 +148,32 @@ func (a *StatefulSetAdapter) ApplyPoolTemplate(ud *alpha1.UnitedDeployment, pool
 	set.Spec.VolumeClaimTemplates = ud.Spec.WorkloadTemplate.StatefulSetTemplate.Spec.VolumeClaimTemplates
 
 	attachNodeAffinityAndTolerations(&set.Spec.Template.Spec, poolConfig)
+
+	if poolConfig.Patches == nil {
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/adapter/statefulset_adapter.go:152]: The code block here is quite similar to the code in deployment adapter. Can you extract them to a common method?

File: pkg/yurtappmanager/controller/uniteddeployment/pool.go [hunk 3]
```diff
@@ -30,6 +30,7 @@ type Pool struct {
 	Namespace string
 	Spec      PoolSpec
 	Status    PoolStatus
+	Patches   string
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/pool.go:33]: How about move "Patch" to Status, called PatchInfo?

File: pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go [hunk 4]
```diff
@@ -97,3 +102,18 @@ func filterOutCondition(conditions []unitv1alpha1.UnitedDeploymentCondition, con
...
 }
+
+func GetNextPatches(ud *unitv1alpha1.UnitedDeployment) map[string]UnitedDeploymentPatches{
+	next := make(map[string]UnitedDeploymentPatches)
+	for _, pool := range ud.Spec.Topology.Pools {
+		t := UnitedDeploymentPatches{}
+		if pool.Replicas != nil {
+			t.Replicas = *pool.Replicas
+		}
+		if pool.Patches != nil {
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go:113]: You may need to think about a problem here. 

If the Patch changes the "replica", the replica value in the UnitedDeploymentPatches does not match the final replica number.

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go [hunk 5]
```diff
@@ -103,3 +108,29 @@ func getCurrentPartition(pods []*corev1.Pod, revision string) *int32 {
...
+		return  err
+	}
+
+	patchedObjMap, err := strategicpatch.StrategicMergeMapPatch(originalObjMap, patchMap, newPatched)
+	if err != nil {
+		klog.Errorf("StartegicMergeMapPatch error %v", err)
+		return err
+	}
+
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(patchedObjMap, newPatched); err != nil {
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go:131]: Why do you need to call this?

**Review by Fei-Guo — COMMENTED** [review]
As a best practice, please update the copyright year to 2021 for all the files that you have touched in the PR.

I think you need to add some UTs to test the merge functionality.

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go (hunk 5 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go:131]: `StrategicMergeMapPatch`  function return a `map[string]interface{}` data.  and use `FromUnstructured` function converts an object from map[string]interface{} representation into a concrete type(Deployment or Statefulset).

and the third argument to function `StrategicMergeMapPatch` is used to determine the type, not for assignment.

File: pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go [hunk 6]
```diff
@@ -124,6 +125,11 @@ type Pool struct {
 	// Indicates the number of the pod to be created under this pool.
 	// +required
 	Replicas *int32 `json:"replicas,omitempty"`
+
+	// Indicates the patches for the templateSpec
+	// Now support strategic merge path :https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch
+	// +optional
+	Patches *runtime.RawExtension `json:"patches,omitempty"`
```
[the remark this answers is not in the document]
Comment (kadisi) [reply pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go:132]: done

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/deployment_adapter.go (hunk 1 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/adapter/deployment_adapter.go:158]: done

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/statefulset_adapter.go (hunk 2 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/adapter/statefulset_adapter.go:152]: done

File: pkg/yurtappmanager/controller/uniteddeployment/pool.go (hunk 3 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/pool.go:33]: done

File: pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go (hunk 4 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go]: `Patch` takes precedence over `Replicas` fields. If the `Patch` also modifies the replicas, use the replicas value in the Patch.

**Review by kadisi — COMMENTED** [review]
done

**Comment by kadisi** [comment]
@Fei-Guo PTAL

**Comment by kadisi** [comment]
> @kadisi Would you make up an example that uses the patch field?

@rambohe-ch done

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go [hunk 7]
```diff
@@ -103,3 +107,56 @@ func getCurrentPartition(pods []*corev1.Pod, revision string) *int32 {
...
+		klog.Errorf("FromUnstructured error %v", err)
+		return err
+	}
+	return nil
+}
+
+func IsPoolHasPatch(poolConfig *appsv1alpha1.Pool, set metav1.Object) bool {
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go:137]: IsPoolHasPatch -> PoolHasPatch

File: pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go [hunk 8]
```diff
@@ -32,6 +32,11 @@ import (
 
 const updateRetries = 5
 
+type UnitedDeploymentPatches struct {
+	Replicas int32
+	Patches  string
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go:37]: Patches -> Patch

File: pkg/yurtappmanager/apis/apps/v1alpha1/well_known_labels_annotations.go [hunk 9]
```diff
@@ -30,6 +30,9 @@ const (
 
 	// SpecifiedDeleteKey indicates this object should be deleted, and the value could be the deletion option.
 	SpecifiedDeleteKey = "apps.openyurt.io/specified-delete"
+
+	// AnnotationPatchesKey indicates the patches for every sub pool
+	AnnotationPatchesKey = "apps.openyurt.io/patches"
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/apis/apps/v1alpha1/well_known_labels_annotations.go:35]: AnnotationPatchesKey = "apps.openyurt.io/patches"
->
AnnotationPatchKey = "apps.openyurt.io/patch"

File: pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go [hunk 10]
```diff
@@ -124,6 +125,13 @@ type Pool struct {
 	// Indicates the number of the pod to be created under this pool.
 	// +required
 	Replicas *int32 `json:"replicas,omitempty"`
+
+	// Indicates the patches for the templateSpec
```
Comment (Fei-Guo) [inline pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go:129]: Patches -> patch

File: config/yurt-app-manager/crd/bases/apps.openyurt.io_uniteddeployments.yaml [hunk 11]
```diff
@@ -127,6 +127,10 @@ spec:
                           distributed across multiple groups of nodes. A pool's nodeSelectorTerm
                           is not allowed to be updated.
                         type: object
+                      patch:
+                        description: Indicates the patches for the templateSpec Now
```
Comment (Fei-Guo) [inline config/yurt-app-manager/crd/bases/apps.openyurt.io_uniteddeployments.yaml:131]: patches -> patch

**Review by Fei-Guo — COMMENTED** [review]
Overall LG. Nits

File: config/yurt-app-manager/crd/bases/apps.openyurt.io_uniteddeployments.yaml (hunk 11 above)
Comment (kadisi) [reply config/yurt-app-manager/crd/bases/apps.openyurt.io_uniteddeployments.yaml:131]: done

File: pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go (hunk 10 above)
Comment (kadisi) [reply pkg/yurtappmanager/apis/apps/v1alpha1/uniteddeployment_types.go:129]: done

File: pkg/yurtappmanager/apis/apps/v1alpha1/well_known_labels_annotations.go (hunk 9 above)
Comment (kadisi) [reply pkg/yurtappmanager/apis/apps/v1alpha1/well_known_labels_annotations.go:35]: done

File: pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go (hunk 7 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/adapter/adapter_util.go:137]: done

File: pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go (hunk 8 above)
Comment (kadisi) [reply pkg/yurtappmanager/controller/uniteddeployment/uniteddeployment_controller_utils.go:37]: done

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**